vllm/v1/worker/tpu_model_runner.py

```diff
@@ -620,7 +644,11 @@ def execute_model(
 
         max_gen_len = selected_token_ids.shape[-1]
         if max_gen_len == 1:
+            # ModelWrapper returns reqs output + padding. Remove padding
+            # here to keep dynamic op outside of main graph.
             valid_sampled_token_ids = selected_token_ids.tolist()
```

Comment:
There's recompilation after warmup here.

Reply:
In the `tolist()` move? How come?

Reply:
Because `tolist` moves data from TPU to CPU, and as `selected_token_ids`'s value is not ready, a graph capture and recompilation will be trigged.

Reply:
oh sure, I thought it would just call mark_step on forced sync and realize post processing graph didn't change after the first inference